# Downloading data

- The data I'm using for this project is the [Street Tree List]('https://data.sfgov.org/City-Infrastructure/Street-Tree-List/tkzw-k3nq') from San Francisco Department of Public Works. I downloaded the data on Nov. 6, 2022. At the time, the data set was last updated on Nov. 6, 2022 as well.
- I created a copy of the data set and named it `original_street_tree_list.csv`. I then put the data set in the `raw-data` folder.


# Explore the data

In [1]:
# import packages
import pandas as pd
import altair as alt

In [3]:
# read csv file
sf_trees_original = pd.read_csv('street_tree_list.csv')
sf_trees_original.head()

TreeID     qLegalStatus                                           qSpecies  \
0  217365  Section 806 (d)  Ceanothus 'Ray Hartman' :: California Lilac 'R...   
1   92771   DPW Maintained              Tristaniopsis laurina :: Swamp Myrtle   
2   23904   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
3   28646   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
4  229807   DPW Maintained                 Jacaranda mimosifolia :: Jacaranda   

           qAddress  SiteOrder                       qSiteInfo PlantType  \
0   707 Rockdale Dr        1.0  Sidewalk: Property side : Yard      Tree   
1   11X Blanken Ave        4.0    Sidewalk: Curb side : Cutout      Tree   
2  1600X Webster St        6.0                 Median : Cutout      Tree   
3  1600X Webster St        7.0                 Median : Cutout      Tree   
4    2560 Bryant St        1.0    Sidewalk: Curb side : Cutout      Tree   

  qCaretaker qCareAssistant               PlantDate  ...        XCoord  \
0    Private            NaN  10/14/2021 12:00:00 AM  ...  5.997488e+06   
1    Private            NaN  10/14/2021 12:00:00 AM  ...  6.011718e+06   
2        DPW            NaN                     NaN  ...  6.003596e+06   
3        DPW            NaN                     NaN  ...  6.003558e+06   
4    Private            NaN                     NaN  ...  6.009700e+06   

         YCoord   Latitude   Longitude  \
0  2.098235e+06  37.741209 -122.451285   
1  2.087394e+06  37.712247 -122.401320   
2  2.114195e+06  37.785380 -122.431304   
3  2.114375e+06  37.785872 -122.431449   
4  2.102427e+06  37.753411 -122.409355   

                                    Location  Fire Prevention Districts  \
0   (37.74120925101712, -122.45128526411095)                        9.0   
1  (37.712246915438215, -122.40132023435935)                       10.0   
2   (37.78537959802679, -122.43130418097743)                       13.0   
3   (37.78587163716589, -122.43144931782685)                       13.0   
4   (37.75341142310638, -122.40935530851043)                        2.0   

   Police Districts Supervisor Districts  Zip Codes  Neighborhoods (old)  
0               7.0                  4.0       59.0                 40.0  
1               3.0                  8.0      309.0                  1.0  
2               9.0                 11.0    29490.0                 13.0  
3               9.0                 11.0    29490.0                 13.0  
4               4.0                  7.0    28859.0                 19.0  

[5 rows x 23 columns]

In [4]:
sf_trees_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   TreeID                     196590 non-null  int64  
 1   qLegalStatus               196533 non-null  object 
 2   qSpecies                   196590 non-null  object 
 3   qAddress                   195097 non-null  object 
 4   SiteOrder                  194796 non-null  float64
 5   qSiteInfo                  196590 non-null  object 
 6   PlantType                  196590 non-null  object 
 7   qCaretaker                 196590 non-null  object 
 8   qCareAssistant             24707 non-null   object 
 9   PlantDate                  70878 non-null   object 
 10  DBH                        153021 non-null  float64
 11  PlotSize                   146229 non-null  object 
 12  PermitNotes                53367 non-null   object 
 13  XCoord                     19

In [5]:
# Copy the original dataframe

sf_trees = sf_trees_original.copy()

## Check duplicates

In [6]:
# check if TreeID is unique
sf_trees['TreeID'].nunique()

196590

In [7]:
len(sf_trees)

196590

Seems like there's no duplicated `TreeID`. There are 196590 thousand trees planted in SF as of Nov.6, 2022. 

In [8]:
# make sure the length of dataframe matches the number of unique IDs

assert len(sf_trees) == sf_trees['TreeID'].nunique()

- There are 196590 TreeID, but only 70878 have a plant date. 
- `PlantDate` should also be a date Dtype.
- The `Zip Codes` are weird. There are some 2, 3 digit numbers.  

## Convert data type

In [9]:
# convert the `PlantDate` Column

sf_trees['PlantDate'] = pd.to_datetime(sf_trees['PlantDate'])


In [10]:
sf_trees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196590 entries, 0 to 196589
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   TreeID                     196590 non-null  int64         
 1   qLegalStatus               196533 non-null  object        
 2   qSpecies                   196590 non-null  object        
 3   qAddress                   195097 non-null  object        
 4   SiteOrder                  194796 non-null  float64       
 5   qSiteInfo                  196590 non-null  object        
 6   PlantType                  196590 non-null  object        
 7   qCaretaker                 196590 non-null  object        
 8   qCareAssistant             24707 non-null   object        
 9   PlantDate                  70878 non-null   datetime64[ns]
 10  DBH                        153021 non-null  float64       
 11  PlotSize                   146229 non-null  object  

In [11]:
sf_trees.head()

TreeID     qLegalStatus                                           qSpecies  \
0  217365  Section 806 (d)  Ceanothus 'Ray Hartman' :: California Lilac 'R...   
1   92771   DPW Maintained              Tristaniopsis laurina :: Swamp Myrtle   
2   23904   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
3   28646   DPW Maintained     Prunus subhirtella 'Pendula' :: Weeping Cherry   
4  229807   DPW Maintained                 Jacaranda mimosifolia :: Jacaranda   

           qAddress  SiteOrder                       qSiteInfo PlantType  \
0   707 Rockdale Dr        1.0  Sidewalk: Property side : Yard      Tree   
1   11X Blanken Ave        4.0    Sidewalk: Curb side : Cutout      Tree   
2  1600X Webster St        6.0                 Median : Cutout      Tree   
3  1600X Webster St        7.0                 Median : Cutout      Tree   
4    2560 Bryant St        1.0    Sidewalk: Curb side : Cutout      Tree   

  qCaretaker qCareAssistant  PlantDate  ...        XCoord        YCoord  \
0    Private            NaN 2021-10-14  ...  5.997488e+06  2.098235e+06   
1    Private            NaN 2021-10-14  ...  6.011718e+06  2.087394e+06   
2        DPW            NaN        NaT  ...  6.003596e+06  2.114195e+06   
3        DPW            NaN        NaT  ...  6.003558e+06  2.114375e+06   
4    Private            NaN        NaT  ...  6.009700e+06  2.102427e+06   

    Latitude   Longitude                                   Location  \
0  37.741209 -122.451285   (37.74120925101712, -122.45128526411095)   
1  37.712247 -122.401320  (37.712246915438215, -122.40132023435935)   
2  37.785380 -122.431304   (37.78537959802679, -122.43130418097743)   
3  37.785872 -122.431449   (37.78587163716589, -122.43144931782685)   
4  37.753411 -122.409355   (37.75341142310638, -122.40935530851043)   

   Fire Prevention Districts  Police Districts Supervisor Districts  \
0                        9.0               7.0                  4.0   
1                       10.0               3.0                  8.0   
2                       13.0               9.0                 11.0   
3                       13.0               9.0                 11.0   
4                        2.0               4.0                  7.0   

   Zip Codes  Neighborhoods (old)  
0       59.0                 40.0  
1      309.0                  1.0  
2    29490.0                 13.0  
3    29490.0                 13.0  
4    28859.0                 19.0  

[5 rows x 23 columns]

In [18]:
sf_trees['PlantDate'].max()

Timestamp('2022-11-12 00:00:00')

In [19]:
sf_trees['PlantDate'].min()

Timestamp('1955-09-19 00:00:00')

So the earliest tree plant date is in 1955. The tree is 67 years old now. 

In [18]:
sf_trees.columns

Index(['TreeID', 'qLegalStatus', 'qSpecies', 'qAddress', 'SiteOrder',
       'qSiteInfo', 'PlantType', 'qCaretaker', 'qCareAssistant', 'PlantDate',
       'DBH', 'PlotSize', 'PermitNotes', 'XCoord', 'YCoord', 'Latitude',
       'Longitude', 'Location', 'Fire Prevention Districts',
       'Police Districts', 'Supervisor Districts', 'Zip Codes',
       'Neighborhoods (old)'],
      dtype='object')

In [12]:
# look at the unique categories of trees

sf_trees['qSpecies'].unique()

array(["Ceanothus 'Ray Hartman' :: California Lilac 'Ray Hartman'",
       'Tristaniopsis laurina :: Swamp Myrtle',
       "Prunus subhirtella 'Pendula' :: Weeping Cherry",
       'Jacaranda mimosifolia :: Jacaranda',
       'Afrocarpus gracilior :: Fern Pine',
       'Magnolia grandiflora :: Southern Magnolia',
       'Photinia fraseri :: Photinia: Chinese photinia',
       'Laurus nobilis :: Sweet Bay: Grecian Laurel',
       'Platanus x hispanica :: Sycamore: London Plane',
       'Eucalyptus sideroxylon :: Red Ironbark',
       'Sequoia sempervirens :: Coast Redwood',
       'Pinus radiata :: Monterey Pine',
       "Arbutus 'Marina' :: Hybrid Strawberry Tree",
       'Pyrus calleryana :: Ornamental Pear',
       'Callistemon viminalis :: Weeping Bottlebrush',
       'Lophostemon confertus :: Brisbane Box',
       'Geijera parviflora :: Australian Willow',
       'Acacia melanoxylon :: Blackwood Acacia',
       'Ginkgo biloba :: Maidenhair Tree',
       "Platanus x hispanica 'Columb

In [13]:
# look at the caretaker of trees

sf_trees['qCaretaker'].unique()

array(['Private', 'DPW', 'Mission Verde', 'DPW for City Agency',
       'Fire Dept', 'Rec/Park', 'Mayor Office of Housing', 'Port',
       'SFUSD', 'CAN', 'Health Dept', 'MTA', 'Public Library', 'PUC',
       'Housing Authority', 'Dept of Real Estate', 'Police Dept',
       'Office of Mayor', 'Purchasing Dept', 'Arts Commission',
       'War Memorial', 'City College', 'Asian Arts Commission',
       'Cleary Bros. Landscape'], dtype=object)

In [17]:
sf_trees_species = sf_trees.groupby(['qSpecies']).count()
sf_trees_species_sorted = sf_trees_species.sort_values(by=['TreeID'], ascending=False)
sf_trees_species_sorted

TreeID  qLegalStatus  \
qSpecies                                                                   
Tree(s) ::                                           11806         11799   
Platanus x hispanica :: Sycamore: London Plane       11739         11739   
Lophostemon confertus :: Brisbane Box                 8904          8899   
Metrosideros excelsa :: New Zealand Xmas Tree         8850          8850   
Tristaniopsis laurina :: Swamp Myrtle                 7448          7448   
...                                                    ...           ...   
Acer tegmentosum :: Manchurian snakebark maple           1             1   
Prunus sargentii :: Sargent Cherry                       1             1   
Prunus sargentii 'Columnaris' :: Sargent Cherry...       1             1   
Prunus persica nectarina :: Flowering Nectarine...       1             1   
x Chiranthofremontia lenzii :: Hybrid Monkey Ha...       1             1   

                                                    qAddress  SiteOrder  \
qSpecies                                                                  
Tree(s) ::                                             10727      11586   
Platanus x hispanica :: Sycamore: London Plane         11702      11724   
Lophostemon confertus :: Brisbane Box                   8870       8879   
Metrosideros excelsa :: New Zealand Xmas Tree           8829       8802   
Tristaniopsis laurina :: Swamp Myrtle                   7437       7345   
...                                                      ...        ...   
Acer tegmentosum :: Manchurian snakebark maple             1          1   
Prunus sargentii :: Sargent Cherry                         1          1   
Prunus sargentii 'Columnaris' :: Sargent Cherry...         1          0   
Prunus persica nectarina :: Flowering Nectarine...         1          1   
x Chiranthofremontia lenzii :: Hybrid Monkey Ha...         1          1   

                                                    qSiteInfo  PlantType  \
qSpecies                                                                   
Tree(s) ::                                              11806      11806   
Platanus x hispanica :: Sycamore: London Plane          11739      11739   
Lophostemon confertus :: Brisbane Box                    8904       8904   
Metrosideros excelsa :: New Zealand Xmas Tree            8850       8850   
Tristaniopsis laurina :: Swamp Myrtle                    7448       7448   
...                                                       ...        ...   
Acer tegmentosum :: Manchurian snakebark maple              1          1   
Prunus sargentii :: Sargent Cherry                          1          1   
Prunus sargentii 'Columnaris' :: Sargent Cherry...          1          1   
Prunus persica nectarina :: Flowering Nectarine...          1          1   
x Chiranthofremontia lenzii :: Hybrid Monkey Ha...          1          1   

                                                    qCaretaker  \
qSpecies                                                         
Tree(s) ::                                               11806   
Platanus x hispanica :: Sycamore: London Plane           11739   
Lophostemon confertus :: Brisbane Box                     8904   
Metrosideros excelsa :: New Zealand Xmas Tree             8850   
Tristaniopsis laurina :: Swamp Myrtle                     7448   
...                                                        ...   
Acer tegmentosum :: Manchurian snakebark maple               1   
Prunus sargentii :: Sargent Cherry                           1   
Prunus sargentii 'Columnaris' :: Sargent Cherry...           1   
Prunus persica nectarina :: Flowering Nectarine...           1   
x Chiranthofremontia lenzii :: Hybrid Monkey Ha...           1   

                                                    qCareAssistant  PlantDate  \
qSpecies                                                                        
Tree(s) ::                                                     353      10183

In [14]:
sf_trees.groupby(['qCaretaker']).count()

TreeID  qLegalStatus  qSpecies  qAddress  SiteOrder  \
qCaretaker                                                                     
Arts Commission              31            31        31        31         30   
Asian Arts Commission         6             6         6         6          6   
CAN                           9             9         9         9          9   
City College                 11            11        11        11         11   
Cleary Bros. Landscape        1             1         1         1          1   
DPW                       28254         28212     28254     28251      27988   
DPW for City Agency         213           213       213       213        212   
Dept of Real Estate          94            94        94        94         94   
Fire Dept                    69            69        69        69         69   
Health Dept                  56            56        56        56         55   
Housing Authority            73            73        73        73         35   
MTA                         109           109       109       109        108   
Mayor Office of Housing      37            37        37        37         37   
Mission Verde                15            15        15        15         15   
Office of Mayor              27            27        27        27         26   
PUC                         261           261       261       261        260   
Police Dept                  50            50        50        50         50   
Port                        727           727       727       727        725   
Private                  164578        164563    164578    163088     163137   
Public Library               37            37        37        37         37   
Purchasing Dept              85            85        85        85         85   
Rec/Park                    722           722       722       722        720   
SFUSD                      1105          1105      1105      1105       1066   
War Memorial                 20            20        20        20         20   

                         qSiteInfo  PlantType  qCareAssistant  PlantDate  \
qCaretaker                                                                 
Arts Commission                 31         31               0         11   
Asian Arts Commission            6          6               1          0   
CAN                              9          9               8          9   
City College                    11         11               2          2   
Cleary Bros. Landscape           1          1               0          0   
DPW                          28254      28254             926      10913   
DPW for City Agency            213        213             103         66   
Dept of Real Estate             94         94               2         43   
Fire Dept                       69         69               2         44   
Health Dept                     56         56               1         15   
Housing Authority               73         73              60         28   
MTA                            109        109              22          9   
Mayor Office of Housing         37         37               0          4   
Mission Verde                   15         15              12         13   
Office of Mayor                 27         27               2         19   
PUC                            261        261               5         86   
Police Dept                     50         50               1         13   
Port                           727        727               6        629   
Private                     164578     164578           22624      58227   
Public Library                  37         37              13         14   
Purchasing Dept                 85         85               0         42   
Rec/Park                       722        722              15        203   
SFUSD                         1105       1105             902        487   
War Memorial                    20         2